src/vs/workbench/contrib/positronNotebook/browser/selectionMachine.ts

```diff
@@ -325,7 +471,79 @@ export class SelectionStateMachine extends Disposable {
 		return cells[cells.length - 1];
 	}
 
+	/**
+	 * Enforces the invariant: NoCells ↔ cells.length === 0
+	 * Called automatically when cells array changes
+	 */
+	private _enforceInvariant(cells: IPositronNotebookCell[]): void {
+		const currentState = this._state.get();
+
+		if (cells.length === 0 && currentState.type !== SelectionState.NoCells) {
+			// Cells disappeared → force NoCells state
+			this._logService.debug('SelectionMachine: Enforcing NoCells state (no cells exist)');
+			this._setState({ type: SelectionState.NoCells });
+		}
+		else if (cells.length > 0 && currentState.type === SelectionState.NoCells) {
+			// Cells appeared → automatically select first cell
+			this._logService.debug('SelectionMachine: Auto-selecting first cell (cells appeared)');
+			this._setState({
+				type: SelectionState.SingleSelection,
+				selected: [cells[0]]
+			});
+		}
+	}
+
+	/**
+	 * Validates whether a transition from one state to another is valid.
+	 * @param from The source state
+	 * @param to The destination state
+	 * @returns True if the transition is valid, false otherwise
+	 */
+	private _isValidTransition(from: SelectionState, to: SelectionState): boolean {
+		// Define valid transitions as a map
+		const validTransitions: Record<SelectionState, SelectionState[]> = {
```

Comment:
Have you found this useful during dev? I wonder if we can extract this map to top and annotate it as part of the spec instead of having both a natural language description and this?